schnorr_keypair.go

```diff
@@ -0,0 +1,165 @@
+package secp256k1
+
+// #include "./depend/secp256k1/include/secp256k1_extrakeys.h"
+// #include "./depend/secp256k1/include/secp256k1_schnorrsig.h"
+import "C"
+import "C"
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"fmt"
+	"github.com/pkg/errors"
+	"unsafe"
+)
+
+// SchnorrKeyPair is a type representing a pair of Secp256k1 private and public keys.
+// This can be used to create Schnorr signatures
+type SchnorrKeyPair struct {
+	keypair C.secp256k1_keypair
+}
+
+// SerializedPrivateKey is a byte array representing the storage representation of a SchnorrKeyPair
+type SerializedPrivateKey [SerializedPrivateKeySize]byte
+
+// String returns the SchnorrKeyPair as the hexadecimal string
+func (key SerializedPrivateKey) String() string {
+	return hex.EncodeToString(key[:])
+}
+
+// String returns the SchnorrKeyPair as the hexadecimal string
+func (key SchnorrKeyPair) String() string {
+	return key.SerializePrivateKey().String()
+}
+
+// DeserializePrivateKey returns a SchnorrKeyPair type from a 32 byte private key.
+// will verify it's a valid private key(Group Order > key > 0)
+func DeserializePrivateKey(data *SerializedPrivateKey) (*SchnorrKeyPair, error) {
+	key := SchnorrKeyPair{}
+	cPtrPrivateKey := (*C.uchar)(&data[0])
+
+	ret := C.secp256k1_keypair_create(context, &key.keypair, cPtrPrivateKey)
+	if ret != 1 {
+		return nil, errors.New("invalid SchnorrKeyPair (zero or bigger than the group order)")
+	}
+	return &key, nil
+}
+
+// DeserializePrivateKeyFromSlice returns a SchnorrKeyPair type from a serialized private key slice.
+// will verify that it's 32 byte and it's a valid private key(Group Order > key > 0)
+func DeserializePrivateKeyFromSlice(data []byte) (key *SchnorrKeyPair, err error) {
+	if len(data) != SerializedPrivateKeySize {
+		return nil, errors.Errorf("invalid private key length got %d, expected %d", len(data),
+			SerializedPrivateKeySize)
+	}
+
+	serializedKey := &SerializedPrivateKey{}
+	copy(serializedKey[:], data)
+	return DeserializePrivateKey(serializedKey)
+}
+
+// GeneratePrivateKey generates a random valid private key from `crypto/rand`
+func GeneratePrivateKey() (key *SchnorrKeyPair, err error) {
+	rawKey := SerializedPrivateKey{}
+	for {
+		n, err := rand.Read(rawKey[:])
+		if err != nil || n != len(rawKey) {
+			return nil, err
+		}
+		key, err = DeserializePrivateKey(&rawKey)
+		if err == nil {
+			return key, nil
+		}
+	}
+}
+
+// SerializePrivateKey returns the private key in the keypair.
+func (key *SchnorrKeyPair) SerializePrivateKey() *SerializedPrivateKey {
+	// TODO: Replace with upstream function when merged: https://github.com/bitcoin-core/secp256k1/pull/845
+	ret := SerializedPrivateKey{}
+	for i := 0; i < 32; i++ {
```

Comment:
Because the array is of type `[]C.uchar` while I want to return a []byte, so I can either use unsafe to convert one to the other and then use copy, or just loop